internal/venice-common/src/main/java/com/linkedin/venice/throttle/TokenBucket.java

```diff
@@ -71,10 +71,10 @@ public TokenBucket(long capacity, long refillAmount, long refillInterval, TimeUn
   }
 
   /**
-   *
-   * @return true if tokens may have been added, false if short circuited and no tokens were added
+   * Check and add tokens if conditions are met. Note that token may have been updated by another thread even when the
+   * function is short-circuited. Consumers of the token bucket should always retry.
    */
-  private boolean update() {
+  private void update() {
     if (clock.millis() > nextUpdateTime) {
```

Comment:
Looks like `nextUpdateTime` is checked but not safely updated inside.  Put this under the sync-block ?

Reply:
Synced offline, this fix should be sufficient but there are other alternative like a read-write lock that could be easier to reason about but maybe less performant. wdyt @gaojieliu